### ISSUE TITLE
Correlation Analysis Issues & Internal Testing Feedback

### ISSUE THREAD

**Issue by neilkakkar** [PostHog/posthog#6278]
Using this to track things to improve with correlation analysis: (cc: @hazzadous )

1. [ ] Number of items to return based on the odds: Decide a lower threshold: 1.1x as likely isn't very useful. Bias towards more useful insights vs too many insights.
2. [ ] Discard random internal events
3. [ ] Discard funnel steps from correlated events
4. [ ] Lower bound for Events search
5. [ ]

**Comment by marcushyett-ph** [comment]
Firstly WOW @neilkakkar so fast! This is really impressive.

Issue 1: High likelyhood but no successes?

<img width="1288" alt="image" src="https://user-images.githubusercontent.com/85295485/136392323-11859bd1-911e-4e03-9464-317df1eaf3a6.png">

https://app.posthog.com/insights?insight=FUNNELS&funnel_viz_type=steps&interval=day&events=%5B%7B%22id%22%3A%22%24pageview%22%2C%22name%22%3A%22%24pageview%22%2C%22type%22%3A%22events%22%2C%22order%22%3A0%2C%22properties%22%3A%5B%7B%22key%22%3A%22%24current_url%22%2C%22value%22%3A%22https%3A%2F%2Fposthog.com%22%2C%22operator%22%3A%22icontains%22%2C%22type%22%3A%22event%22%7D%5D%7D%2C%7B%22id%22%3A%22user%20signed%20up%22%2C%22name%22%3A%22user%20signed%20up%22%2C%22type%22%3A%22events%22%2C%22order%22%3A1%7D%5D&funnel_from_step=0&funnel_to_step=1&actions=%5B%5D&new_entity=%5B%5D&date_from=-30d

2. Spurious items

It seems like items with very low numbers of events should also be excluded as well as ones with low odds:

<img width="1295" alt="image" src="https://user-images.githubusercontent.com/85295485/136393597-0cd02342-4d54-4295-92c8-73dc187136fd.png">

Some of these events look promising but the numbers are too low to be meaningful


3. Question: Which events do we count? 
- Do we count events before the last event and after the first event - or do we include others?


4. Actions?
- I'm sure this is going to be a pain - but I think supporting actions is going to be a valuable way to improve the signal to noise ratio.

**Comment by marcushyett-ph** [comment]
Also - perhaps there's something weird with the instrumentation of this event:

<img width="474" alt="image" src="https://user-images.githubusercontent.com/85295485/136400582-9fd793f3-d16a-4a61-9126-70f6a026fa01.png">

But it looks like if you pay your bill, you're much less likely to be successful at anything.

**Comment by marcushyett-ph** [comment]
I admit this is very weird way to use this tool but found some great insights by creating a funnel that is looking for people who've logged in 5 times (assuming these folks are retaining well)

But it highlights how much value people get from:

- Viewing person profiles
- Configuring filters
- Creating and updating actions
- Dashboards

https://app.posthog.com/insights?insight=FUNNELS&filter_test_accounts=true&events=%5B%5D&actions=%5B%7B%22id%22%3A%225043%22%2C%22name%22%3A%22App%20Pageview%20-%20Logged%20in%22%2C%22type%22%3A%22actions%22%2C%22order%22%3A0%7D%2C%7B%22id%22%3A%225043%22%2C%22name%22%3A%22App%20Pageview%20-%20Logged%20in%22%2C%22type%22%3A%22actions%22%2C%22order%22%3A1%7D%2C%7B%22id%22%3A%225043%22%2C%22name%22%3A%22App%20Pageview%20-%20Logged%20in%22%2C%22type%22%3A%22actions%22%2C%22order%22%3A2%7D%2C%7B%22id%22%3A%225043%22%2C%22name%22%3A%22App%20Pageview%20-%20Logged%20in%22%2C%22type%22%3A%22actions%22%2C%22order%22%3A3%7D%2C%7B%22id%22%3A%225043%22%2C%22name%22%3A%22App%20Pageview%20-%20Logged%20in%22%2C%22type%22%3A%22actions%22%2C%22order%22%3A4%7D%5D&funnel_viz_type=steps&display=FunnelViz&interval=day&funnel_from_step=0&funnel_to_step=1&properties=%5B%5D&new_entity=%5B%5D&date_from=-90d


Also 1 final question @neilkakkar the query runs super super fast - I thought you have speed / scalability concerns before - what's changed?

**Comment by neilkakkar** [comment]
Wow this is SO clever! Haha, noice.

regd: query times:

1. We halved execution time already by backend optimisations.
2. **It's slower than you think**: I just don't remove old data while it's getting new data right now (rough MVP)
3. We'll optimise this further, since right now the query is triggered only after funnels results have come, vs, in parallel (rough MVP)

A lot of slowness was also due to running queries one by one per property, which has also now been optimized away (to come soon)

**Comment by paolodamico** [comment]
Incredibly @neilkakkar, moving faaast! Played around with it and some feedback,

- Tested [this funnel](https://app.posthog.com/insights?insight=FUNNELS&display=FunnelViz&events=%5B%7B%22id%22%3A%22first%20team%20event%20ingested%22%2C%22type%22%3A%22events%22%2C%22order%22%3A0%2C%22name%22%3A%22first%20team%20event%20ingested%22%2C%22custom_name%22%3Anull%2C%22math%22%3Anull%2C%22math_property%22%3Anull%2C%22properties%22%3A%5B%5D%7D%5D&actions=%5B%7B%22id%22%3A4912%2C%22type%22%3A%22actions%22%2C%22order%22%3A1%2C%22name%22%3A%22Analyzed%20Data%2C%20core%20analytics%22%2C%22custom_name%22%3Anull%2C%22math%22%3Anull%2C%22math_property%22%3Anull%2C%22properties%22%3A%5B%5D%7D%2C%7B%22id%22%3A%2210784%22%2C%22type%22%3A%22actions%22%2C%22order%22%3A2%2C%22name%22%3A%22Product%20Metric%20%5BName%20TBD%5D%22%2C%22math%22%3Anull%2C%22math_property%22%3Anull%2C%22properties%22%3A%5B%5D%7D%5D&funnel_from_step=0&funnel_to_step=1&funnel_viz_type=steps&funnel_window_days=14&interval=day&properties=%5B%7B%22key%22%3A%22id%22%2C%22value%22%3A71%2C%22operator%22%3Anull%2C%22type%22%3A%22cohort%22%7D%5D&new_entity=%5B%5D&date_from=-90d) which ends with the "Discoveries" action. This action only has an underlying event of "insight analyzed" and I'm seeing it as a suggestion. Ideally we would remove any events that comprise the action we're targeting. Also, seems quite strange to have  a drop-off here.
    <img width="1000" alt="" src="https://user-images.githubusercontent.com/5864173/136498640-80cb1931-e8f3-4731-ab04-95c6754cf94b.png">
- Maybe we should consider removing results that aren't statistically significant? For instance I have a 13x more likely to convert but only 3 users converted.
- How do we measure the user read these insights? Do we have some custom events here? Should we maybe trigger a custom event when you've scrolled the table into view and stay there for >10s?
- More of a general comment that something actionable, but I don't think I get a ton of value from this if I haven't set up a ton of relevant custom events. Keen on trying user properties to see how these better cover this gap because it's a reality that a lot of our users won't have a ton of events properly set up.

**Comment by neilkakkar** [comment]
### Question: Which events do we count?

We count all events between start of funnel and successful conversion

### Actions?

You got that right, will defo be a pain, but also good point about actions being good signals: they'd rarely be fluff events. Not going to push this into the MVP for now 😅 .

---

Made some updates, all the above results have changed! (For the better I think). (If your funnel looks the same, try changing the conversion window from 14 to 15 - we have caching issues with these results).

It's not there on the frontend yet, but now the API returns a skewed parameter, which tells you when the input data is too skewed to produce meaningful results. Specially valid [for this funnel](https://app.posthog.com/insights?insight=FUNNELS&funnel_viz_type=steps&interval=day&events=%5B%7B%22id%22%3A%22%24pageview%22%2C%22name%22%3A%22%24pageview%22%2C%22type%22%3A%22events%22%2C%22order%22%3A0%2C%22properties%22%3A%5B%7B%22key%22%3A%22%24current_url%22%2C%22value%22%3A%22https%3A%2F%2Fposthog.com%22%2C%22operator%22%3A%22icontains%22%2C%22type%22%3A%22event%22%7D%5D%7D%2C%7B%22id%22%3A%22user%20signed%20up%22%2C%22name%22%3A%22user%20signed%20up%22%2C%22type%22%3A%22events%22%2C%22order%22%3A1%7D%5D&funnel_from_step=0&funnel_to_step=1&actions=%5B%5D&new_entity=%5B%5D&date_from=-30d&funnel_window_interval_unit=day&funnel_window_interval=13) - too few successes that even a slight signal boosts it up.

I'm planning to add a warning for this. Or, optionally error out / figure out another way around.

---

Paolo's example funnel was great for no-skew. I like how the results look now:
![image](https://user-images.githubusercontent.com/7115141/136554298-5466d0d0-fc9e-4cfd-bab7-2a9f3375d397.png)

Obviously, the internal events seem not very meaningful. Planning on removing them, but not yet, want to see them consistently be useless first.

---

About how users read these insights, yep, no event yet! :$ What you say makes sense - not sure about the time though? 10s feels like a long time

---

And properties should be up within the hour too :)

**Comment by neilkakkar** [comment]
> But it looks like if you pay your bill, you're much less likely to be successful at anything

Did a small deep dive on this on Friday: It's indeed instrumentation. From slack: https://posthog.slack.com/archives/CSPHFDZH8/p1633708858316400?thread_ts=1633701275.300800&cid=CSPHFDZH8

The billing period is correlated with sign up date, hence event sometimes coincides at the same time. The event is sent as a background task, after some checks, etc. - and not related to when the user is doing the event, from what I understand.

**Comment by marcushyett-ph** [comment]
Great thanks for the research - we should have a way for an team / project to filter out spurious events like this - so they other people don't get caught out by them.

**Comment by neilkakkar** [comment]
Pros and Cons of running correlation on everything:

Pro: You find really cool stuff you didn't think of checking, like finding out which organisations are most engaged.

 User Signed Up -> Insight Analyzed funnel : 
![image](https://user-images.githubusercontent.com/7115141/136786294-f82c9bdd-a8db-4e8e-ad26-1859a4345327.png)

Con: You get a lot of duplicates - properties that are correlated between themselves show up together. This is okay if you have enough results, but not if you only have, say 3-4 results, all filled with the correlated properties.

You can tell how they always show up together, by checking the success/dropoff numbers to be the same 😅

**Comment by marcushyett-ph** [comment]
Wow this is really really cool.

> You can tell how they always show up together, by checking the success/dropoff numbers to be the same

If this is the case - shouldn't these be really easy to filter out?

**Comment by neilkakkar** [comment]
It's a simple heuristic :$ - not really sure if it will hold, need more testing. And then there's the question of choosing which one out of the many in the same group is meaningful.

Choosing project_count = 18, vs the organization_id = bla bla> - makes a tonne of difference.  The former is meaningless, while the latter is very specific.

(Also, if there were another organization with project_count = 18, that would've perturbed the result for project_count. Along similar lines, something else could get the same counts while not being correlated to a property, which can result in us discarding valuable information)

**Comment by neilkakkar** [comment]
> we should have a way for an team / project to filter out spurious events like this - so they other people don't get caught out by them.

Agree.

Excluding properties / events I know to be useless can be a really powerful complement to the above with $all properties as well.

And as @paolodamico mentioned https://github.com/PostHog/product-internal/issues/178#issuecomment-939981303.

**Comment by marcushyett-ph** [comment]
> It's a simple heuristic :$ - not really sure if it will hold, need more testing

Cool. Looking at a bunch of the things that have >100x correlations - they're all pretty meaningless. Perhaps a simple cap on these "overfitted" properties would solve.

In my mind the best experience is going to be showing people a few "nuggets" which we have high confidence in and if that returns nothing or they want to explore further they can click through and review everything.

**Comment by neilkakkar** [comment]
What properties / events were these?

**Comment by jamesefhawkins** [comment]
Sall piece of feedback - got a bit confused by terminology:

Which "event" does this refer to? Should it say "90x less likely to convert" to make clear it means getting throught the entire funnel? Secondly, I wasn't sure if "this event' could mean "completed" OR "dropped off".
<img width="435" alt="Screenshot 2021-10-11 at 14 27 42" src="https://user-images.githubusercontent.com/47497682/136798184-ed51de2a-9488-46a2-936f-b38f46cecd70.png">

**Comment by marcushyett-ph** [comment]
@neilkakkar basically the entire list for $all in [this funnel](https://app.posthog.com/insights?insight=FUNNELS&properties=%5B%5D&filter_test_accounts=true&events=%5B%7B%22id%22%3A%22%24pageview%22%2C%22name%22%3A%22%24pageview%22%2C%22type%22%3A%22events%22%2C%22order%22%3A0%2C%22math%22%3A%22dau%22%2C%22properties%22%3A%5B%7B%22key%22%3A%22%24current_url%22%2C%22value%22%3A%22https%3A%2F%2Fposthog.com%22%2C%22operator%22%3A%22icontains%22%2C%22type%22%3A%22event%22%7D%5D%7D%2C%7B%22id%22%3A%22user%20signed%20up%22%2C%22name%22%3A%22user%20signed%20up%22%2C%22type%22%3A%22events%22%2C%22order%22%3A1%2C%22math%22%3A%22dau%22%2C%22properties%22%3A%5B%5D%7D%5D&actions=%5B%5D&funnel_viz_type=steps&display=FunnelViz&interval=day&funnel_from_step=0&funnel_to_step=1&new_entity=%5B%5D#fromItem=451044)

**Comment by neilkakkar** [comment]
@jamesefhawkins : What it tries to say is:

People who completed the funnel were 90x less likely to do the "user instance status report" event. Or put another way:
People who dropped off were 90x more likely to do the "user instance status report" event.

It doesn't try to establish causality. Aka **it doesn't try to say** "If people do the 'user instance status report' event, they'll be 90x less likely to convert.

Was this not so clear?

**Comment by neilkakkar** [comment]
Most internal issues have now been sorted, results look reasonably good. Will open a new issue based on feedback from customers.
